src/app/app_execution.rs

```diff
@@ -0,0 +1,110 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! [`AppExecution`]: Handles executing queries for the TUI application.
+
+use crate::app::state::tabs::sql::Query;
+use crate::app::AppEvent;
+use crate::execution::ExecutionContext;
+use color_eyre::eyre::Result;
+use futures::StreamExt;
+use log::{error, info};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::mpsc::UnboundedSender;
+
+/// Handles executing queries for the TUI application, formatting results
+/// and sending them to the UI.
+pub(crate) struct AppExecution {
+    inner: Arc<ExecutionContext>,
+}
+
+impl AppExecution {
+    /// Create a new instance of [`AppExecution`].
+    pub fn new(inner: Arc<ExecutionContext>) -> Self {
+        Self { inner }
+    }
+
+    /// Run the sequence of SQL queries, sending the results as [`AppEvent::QueryResult`] via the sender.
+    ///
+    /// All queries except the last one will have their results discarded.
+    ///
+    /// Error handling: If an error occurs while executing a query, the error is
+    /// logged and execution continues
+    pub async fn run_sqls(&self, sqls: Vec<&str>, sender: UnboundedSender<AppEvent>) -> Result<()> {
```

Comment:
I moved the code that executes queries and sends them as AppEvents into its own struct